content/docs/reference/design-docs/editoast-errors/_index.en.md

```diff
@@ -0,0 +1,320 @@
+    )
+)]
+async fn endpoint(Path(key): Path<Key>) -> Result<Json<Computation>, EndpointError> {
+    let resource = get_resource(key)?;
+    let value = process_resource(resource)?;
+    Ok(Json(value))
+}
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+pub enum EndpointError {
+    #[error("Resource not found")]
+    #[editoast_error(user)] // <=> status = 400
+    ResourceNotFound(#[from] GetError),
+
+    #[editoast_error(internal)] // <=> status = 500, default, overrides the message to "Internal error - please contact your administrator"
+    ProcessingFailed(#[from] ProcessingError)
+}
+```
+
+## Share errors between crates
+
+Since we require no other constraint that `impl Error` for composition, it's easy to nest errors using `thiserror`.
+
+````rust
+// in editoast_models
+
+#[derive(Debug, thiserror::Error)]
+#[error("postgres error: {0}")]
+struct DbError(#[from] diesel::Error);
+
+// in editoast_redis (if we actually had that crate 👀)
+
+#[derive(Debug, thiserror::Error)]
+#[error("redis error: {0}")]
+struct RedisError(#[from] redis::Error);
+
+// in editoast_views, where diesel isn't available
+
+#[derive(Debug, thiserror::Error)]
+#[error("invalid resource form: {0}")]
+struct FormError(ResourceForm);
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+#[editoast_error(name = "CreateResourceError")] // schema name & error_type
+enum CreateError {
+    #[error("will be overriden, but still useful for development")]
+    #[editoast_error(internal, name = "Database")]
+    Db(#[from] DbError),
+
+    #[error(transparent)] // shown to the user
+    #[editoast_error(user)]
+    InvalidForm(#[from] FormError)
+}
+
+async fn create(...) -> Result<Json<Resource>, CreateError> {
+    todo!()
+}
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+enum UpdateError {
+    #[editoast_error(internal)]
+    Db(#[from] DbError),
+
+    #[editoast_error(internal)]
+    Redis(#[from] RedisError),
+
+    #[error(transparent)]
+    #[editoast_error(user)]
+    InvalidForm(#[from] FormError)
+}
+
+async fn update(...) -> Result<(), UpdateError> {
+    todo!()
+}
+````
+
+## Ease composability
+
+### Nesting `EditoastError`s
+
+We composed errors by nesting them thanks to `thiserror`. However, to compose and reuse `EditoastErrors`, we need a special flag so that when we attempt to serialize the error, we return the serialization of the source error directly.
+
+```rust
+#[derive(Debug, thiserror::Error, EditoastError)]
+#[error("no such infra: {id}")]
+#[editoast_error(context, status = 404)]
+struct InfraNotFound { id: u64 }
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+#[error("unauthorized")]
+#[editoast_error(status = 401)]
+struct Unauthorized;
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+enum EndpointError {
+    NotFound(#[from] #[editoast_error] InfraNotFound),
+
+    Unauthorized(#[from] #[editoast_error] Unauthorized)
+
+    #[error("oh no")]
+    #[editoast_error(user)]
+    Error1,
+}
+```
+
+### Anonymous enums
```

Comment:
I fully agree. Writing error enum isn't fun, but with thiserror it isn't that cumbersome, and has the nice side effect of being usable without refering to same `T2` variant...

Reply:
The point was to find a way to use them transparently by just using the implicit `from()` that goes with the `?` operator. But after prototyping a bit I couldn't find a way to make that work with rust stable. `anon_enum` not implementing `From` either seems to confirm that. I kept the section with an explanation block on why we won't do it, just in case some Rust wizard comes by and magically find a solution... Cf. [`717e31d` (#237)](https://github.com/OpenRailAssociation/osrd-website/pull/237/commits/717e31df20a3a87ae897ed7212f155b6fa3d5e72)